src/Qir/Runtime/lib/Simulators/FullstateSimulator.cpp

```diff
@@ -238,6 +261,11 @@ namespace Quantum
             typedef unsigned (*TMeasure)(unsigned, unsigned, unsigned*, unsigned*);
             static TMeasure m         = reinterpret_cast<TMeasure>(this->GetProc("Measure"));
             std::vector<unsigned> ids = GetQubitIds(numTargets, targets);
+            if (ids.size() == 1)
+            {
+                // If measuring exactly one qubit, mark it as measured for tracking.
+                isMeasured[ids[0]] = true;
+            }
             return reinterpret_cast<Result>(
                 m(this->simulatorId, (unsigned)numBases, reinterpret_cast<unsigned*>(bases), ids.data()));
         }

```

Comment:
I got an impression that the measurement does not have to be in PauliZ basis, to deallocate the qubit correctly. 

After the measurement in PauliX basis the qubit state collapses to one of the values in PauliX basis (which I interpret as equally probable |0> or |1> in PauliZ basis). And such a qubit can still be correctly deallocated (because it has been measured).

Feel free to correct me if I'm wrong.

Reply:
If the result of `Measure([PauliX], [q])` is `Zero`, then the state of `q` is known to be |+⟩ ≔ (|0⟩ + |1⟩) / √2, while if the result is `One`, then `q` must be in |−⟩ ≔ (|0⟩ − |1⟩) / √2. Either way, the state of `q` is known precisely and thus can be efficiently reset to the |0⟩ state without additional measurements (in particular, by `H(q);` in the first case, and by `Z(q); H(q);` in the second). A similar argument applies when measuring in the `PauliY` basis.